### PR TITLE
Add index for VPC store

### PR DESCRIPTION
**Pull request by wenqiq**
Add index for VPC store to support searching VPCs based on Namespace name/id.

### CODE REVIEW

File: pkg/nsx/services/vpc/store.go [hunk 1]
```diff
@@ -30,18 +30,36 @@ func indexFunc(obj interface{}) ([]string, error) {
...
 	case *model.LBService:
-		return filterTag(o.Tags), nil
+		return filterTag(o.Tags, common.TagScopeNamespaceUID), nil
 	default:
 		return res, errors.New("indexFunc doesn't support unknown type")
 	}
 }
 
-var filterTag = func(v []model.Tag) []string {
+func vpcIndexNamespaceFunc(obj interface{}) ([]string, error) {
```
Comment (wenyingd) [inline pkg/nsx/services/vpc/store.go:41]: Can we rename it as `vpcIndexNamespaceNameFunc`

File: pkg/nsx/services/common/store.go [hunk 2]
```diff
@@ -182,7 +180,7 @@ func (service *Service) SearchResource(resourceTypeValue string, queryParam stri
 // PopulateResourcetoStore is the method used by populating resources created not by nsx-operator
 func (service *Service) PopulateResourcetoStore(wg *sync.WaitGroup, fatalErrors chan error, resourceTypeValue string, queryParam string, store Store, filter Filter) {
 	defer wg.Done()
-	count, err := service.SearchResource("", queryParam, store, filter)
+	count, err := service.SearchResource(resourceTypeValue, queryParam, store, filter)
```
Comment (wenyingd) [inline pkg/nsx/services/common/store.go:183]: Is this change related?

File: pkg/nsx/services/vpc/store.go [hunk 3]
```diff
@@ -30,18 +30,36 @@ func indexFunc(obj interface{}) ([]string, error) {
...
+func vpcIndexNamespaceIDFunc(obj interface{}) ([]string, error) {
+	switch o := obj.(type) {
+	case *model.Vpc:
+		return filterTag(o.Tags, common.TagScopeNamespaceUID), nil
+	default:
+		return nil, errors.New("VPCIndexNamespaceFunc doesn't support unknown type")
+	}
+}
+
+var filterTag = func(v []model.Tag, tagScope string) []string {
```
Comment (wenyingd) [inline pkg/nsx/services/vpc/store.go:59]: rename as `filterTagBy`?

File: pkg/nsx/services/common/store.go (hunk 2 above)
Comment (wenqiq) [reply pkg/nsx/services/common/store.go]: Not related, just changed it when I saw it.

File: pkg/nsx/services/vpc/store.go (hunk 3 above)
Comment (wenqiq) [reply pkg/nsx/services/vpc/store.go]: updated.

File: pkg/nsx/services/vpc/vpc_test.go [hunk 4]
```diff
@@ -37,16 +37,17 @@ import (
...
+	vpcName2             = "ns1-vpc-2"
+	infraVPCName         = "infra-vpc"
+	vpcID1               = "ns-vpc-uid-1"
+	vpcID2               = "ns-vpc-uid-2"
+	vpcID3               = "ns-vpc-uid-3"
+	IPv4Type             = "IPv4"
+	cluster              = "k8scl-one"
+	tagScopeCluster      = common.TagScopeCluster
+	tagScopeNamespace    = common.TagScopeNamespace
+	tagScopeNamespaceUID = common.TagScopeNamespaceUID
```
Comment (timdengyun) [inline pkg/nsx/services/vpc/vpc_test.go:50]: There is an error at code line#L63
cache.Indexers{common.TagScopeStaticRouteCRUID: indexFunc}),
```
vpcStore := &VPCStore{ResourceStore: common.ResourceStore{
		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeStaticRouteCRUID: indexFunc}),
		BindingType: model.VpcBindingType(),
```
}}

File: pkg/nsx/services/vpc/store.go [hunk 5]
```diff
@@ -30,18 +30,36 @@ func indexFunc(obj interface{}) ([]string, error) {
...
+func vpcIndexNamespaceNameFunc(obj interface{}) ([]string, error) {
+	switch o := obj.(type) {
+	case *model.Vpc:
+		return filterTagBy(o.Tags, common.TagScopeNamespace), nil
+	default:
+		return nil, errors.New("VPCIndexNamespaceFunc doesn't support unknown type")
+	}
+}
+
+func vpcIndexNamespaceIDFunc(obj interface{}) ([]string, error) {
```
Comment (timdengyun) [inline pkg/nsx/services/vpc/store.go:50]: `vpcIndexNamespaceIDFunc` is the same with `indexFunc`, but we don't see usage for indexFunc`. It's better to check with @seanpang-vmware  and @TaoZou1 
There is no reason to keep duplicated funcs here.

File: pkg/nsx/services/vpc/vpc_test.go (hunk 4 above)
Comment (wenqiq) [reply pkg/nsx/services/vpc/vpc_test.go:50]: line63? Could you make a more clear code suggestion?

File: pkg/nsx/services/vpc/store.go (hunk 5 above)
Comment (wenqiq) [reply pkg/nsx/services/vpc/store.go:36]: This function `indexFunc` is only used in test files and is unnecessary. I deleted it.

**Review by wenyingd — APPROVED** [review]
LGTM

**Review by TaoZou1 — APPROVED** [review]

File: pkg/nsx/services/vpc/store.go [hunk 6]
```diff
@@ -74,28 +78,25 @@ func (vs *VPCStore) Apply(i interface{}) error {
...
-				ret = append(ret, mvpc)
-			}
-		}
-	}
-	return ret
+func (s *VPCStore) GetVPCsByNamespace(ns string) []*model.Vpc {
+	return s.GetByIndex(common.TagScopeNamespace, ns)
+}
+
+func (s *VPCStore) GetVPCsByNamespaceID(namespaceID string) []*model.Vpc {
```
Comment (timdengyun) [inline pkg/nsx/services/vpc/store.go:94]: I'm doubting the use case for  `GetVPCsByNamespaceID`, since usually in fetching stale VPCs case, we need to list all VPS associated with ns name not ns ID. but it's not be harmful to put here at this moment.

**Review by timdengyun — APPROVED** [review]
one doute comment, but ok for now.

File: pkg/nsx/services/vpc/store.go (hunk 6 above)
Comment (wenqiq) [reply pkg/nsx/services/vpc/store.go:94]: Is it related to this PR change? The `GetVPCsByNamespaceID` is used in creating a Subnet, since the VPCs fetched by Namespace ID must not be stale.

Comment (wenqiq) [reply pkg/nsx/services/vpc/store.go:94]: The VPCs fetched by Namespace name may be stale.
